src/libraries/System.Linq/tests/TakeTests.cs

```diff
@@ -544,24 +1271,695 @@ public void ElementAtOfLazySkipTakeChain(IEnumerable<int> source, int skip, int
         [InlineData(2, 3)]
         public void DisposeSource(int sourceCount, int count)
         {
-            int state = 0;
+            int[] state = new int[5];
 
-            var source = new DelegateIterator<int>(
-                moveNext: () => ++state <= sourceCount,
+            var source = Repeat(index => new DelegateIterator<int>(
```

Comment:
This is the only test I could find where `Repeat` is really needed. 

Nevertheless the test has too many moving parts and each section is applying different sets of assertions. I would recommend either breaking this up into separate unit tests or perhaps it might be possible to factor out some of the state management into a local method.

Reply:
Refactored the code.